File: action_sample/src/DoDishes_client.cpp
#include <actionlib/client/simple_action_client.h>
#include "action_sample/DoDishesAction.h"

void doneCb(const actionlib::SimpleClientGoalState& state,
            const action_sample::DoDishesResultConstPtr& result) {
  ROS_INFO("%d dishes is clean, state is %s", result->total_dishes_cleaned, state.toString().c_str());
  ros::shutdown();
}

void activeCb() {
  ROS_INFO("job is active");
}

void feedbackCb(const action_sample::DoDishesFeedbackConstPtr& feedback) {
  ROS_INFO("percent is %f", feedback->percent_complete);
}


int main(int argc, char** argv) {
  ros::init(argc, argv, "do_dishes_client");

  actionlib::SimpleActionClient<action_sample::DoDishesAction> client("do_dishes", true);

  ROS_INFO("waiting for action server to start.");
  client.waitForServer();
  ROS_INFO("action server started, sending goal.");

  action_sample::DoDishesGoal goal;
  goal.dishwasher_id = 1;

  client.sendGoal(goal, &doneCb, &activeCb, &feedbackCb);

  // sleep(4);
  // client.cancelGoal();

  ros::spin();

  return 0;
}

File: action_sample/scripts/DoDishes_client.py
#! /usr/bin/env python3

import rospy
import actionlib

from action_sample.msg import DoDishesAction, DoDishesGoal

def done_cb(state, result):
  rospy.loginfo("%d dishes is clean, state is %s" %(result.total_dishes_cleaned, str(state)))
  rospy.signal_shutdown('Normal shutdown')

def active_cb():
  rospy.loginfo("job is action")
  
def feedback_cb(feedback):
  rospy.loginfo("percent is %f" %feedback.percent_complete)

def main():
  rospy.init_node("do_dishes_client")
  
  client = actionlib.SimpleActionClient("do_dishes", DoDishesAction)
  
  rospy.loginfo("waiting for action server to start")
  client.wait_for_server()
  
  rospy.loginfo("action server started, sending goal")
  goal = DoDishesGoal()
  goal.dishwasher_id = 1

  client.send_goal(goal, done_cb=done_cb, active_cb=active_cb, feedback_cb=feedback_cb)
  
  rospy.sleep(4)
  client.cancel_goal()
  
  rospy.spin()


if __name__ == "__main__":
  main()